### PR TITLE
 Castle.Windsor.Extensions.DependencyInjection: support parallel containers (see #563)

### PR DESCRIPTION
**Pull request by rvdginste**
An attempt to fix the issue in #563.

### CODE REVIEW

File: src/Castle.Windsor.Extensions.DependencyInjection/Scope/ExtensionContainerScope.cs
Comment (generik0) [inline src/Castle.Windsor.Extensions.DependencyInjection/Scope/ExtensionContainerScope.cs:38]: Castle does not want c++ notations, must `if(parent!=null)`

Comment (jonorossi) [reply src/Castle.Windsor.Extensions.DependencyInjection/Scope/ExtensionContainerScope.cs:38]: That is not true, C#'s null coalescing operator is perfectly acceptable. If Castle's coding style guide is confusing, we should update it.

File: src/Castle.Windsor.Extensions.DependencyInjection/Scope/ExtensionContainerScope.cs
Comment (generik0) [inline src/Castle.Windsor.Extensions.DependencyInjection/Scope/ExtensionContainerScope.cs:43]: is "myRootScope" a good name?

Do to Windsor code style you might need to use a traditional getter with if statements

**Comment by generik0** [comment]
@rvdginste @jonorossi 
Reviewed. Looks great, thanks. A couple of code styling changes might be needed. Unless we are allowed to use c++/C# newest syntax now @jonorossi (I could not find the doc file with the code style).

@rvdginste can you give me access to the fork? I have a test to add for this change that tests the muli assembly issues easily.

File: src/Castle.Windsor.Extensions.DependencyInjection/Scope/ExtensionContainerScope.cs
Comment (generik0) [reply src/Castle.Windsor.Extensions.DependencyInjection/Scope/ExtensionContainerScope.cs:38]: Good to know. My misunderstanding of https://github.com/castleproject/Home/blob/master/coding-standards.md#c-style-code

File: src/Castle.Windsor.Extensions.DependencyInjection/Scope/ExtensionContainerScope.cs
Comment (rvdginste) [reply src/Castle.Windsor.Extensions.DependencyInjection/Scope/ExtensionContainerScope.cs]: I agree. I changed it to self, I hope that's better.

**Comment by rvdginste** [comment]
> @rvdginste can you give me access to the fork? I have a test to add for this change that tests the muli assembly issues easily.

@generik0 I sent you an invite.. not clear whether I still have to give you write access though, just let me know.

**Review by generik0 — APPROVED** [review]

**Comment by generik0** [comment]
@rvdginste good work. 
I still don't know why I didn't catch the scope factory, as of cause we should control from the factory...! But sometimes one just goes down the wrong track. Thanks for getting the change back on track 👍 
@jonorossi I have added uTest that used to fail before due to the root scope being statically bound to the test assembly, but now it is bound to the provider factory instance. Much better.
The test also documents how you can add your own container to the mix, allow for extra control and registrations.

File: src/Castle.Windsor.Extensions.DependencyInjection/Scope/ExtensionContainerScope.cs
Comment (rvdginste) [reply src/Castle.Windsor.Extensions.DependencyInjection/Scope/ExtensionContainerScope.cs]: I changed it to your last suggestion, is better.

File: src/Castle.Windsor.Extensions.DependencyInjection/WindsorScopeFactory.cs
Comment (jonorossi) [inline src/Castle.Windsor.Extensions.DependencyInjection/WindsorScopeFactory.cs:44]: Any reason this need 5 lines rather than just 1, it isn't very long?

```csharp
var scope = ExtensionContainerScope.BeginScope(ExtensionContainerScope.Current, rootScope);
```

Comment (jonorossi) [inline src/Castle.Windsor.Extensions.DependencyInjection/WindsorScopeFactory.cs:44]: Does this class need to hold on to the `rootScope`, couldn't it just call `ExtensionContainerScope.Current.RootScope`. Why can't `BeginScope` just get the root scope from the passed scope?

File: src/Castle.Windsor.Extensions.DependencyInjection/Scope/ExtensionContainerRootScopeAccessor.cs
Comment (jonorossi) [inline src/Castle.Windsor.Extensions.DependencyInjection/Scope/ExtensionContainerRootScopeAccessor.cs:37]: This function is probably better written like below, or maybe even more concise, at minimum combine the duplication of if-statements with the same body:

```csharp
public ILifetimeScope GetScope(CreationContext context)
{
    return ExtensionContainerScope.Current?.RootScope ??
        throw new InvalidOperationException("No root scope");
}
```

**Review by jonorossi — CHANGES_REQUESTED** [review]

**Review by jonorossi — CHANGES_REQUESTED** [review]
Thanks @rvdginste. I've included some review comments. Could you please also update the changelog.

File: src/Castle.Windsor.Extensions.DependencyInjection/Scope/ExtensionContainerRootScopeAccessor.cs
Comment (generik0) [reply src/Castle.Windsor.Extensions.DependencyInjection/Scope/ExtensionContainerRootScopeAccessor.cs]: fixed

File: src/Castle.Windsor.Extensions.DependencyInjection/WindsorScopeFactory.cs
Comment (generik0) [reply src/Castle.Windsor.Extensions.DependencyInjection/WindsorScopeFactory.cs]: fixed

Comment (generik0) [reply src/Castle.Windsor.Extensions.DependencyInjection/WindsorScopeFactory.cs]: fixed

File: src/Castle.Windsor.Extensions.DependencyInjection/WindsorScopeFactory.cs
Comment (jonorossi) [inline src/Castle.Windsor.Extensions.DependencyInjection/WindsorScopeFactory.cs:28]: Field `rootScope` is unused.

File: src/Castle.Windsor.Extensions.DependencyInjection/WindsorScopeFactory.cs
Comment (jonorossi) [inline src/Castle.Windsor.Extensions.DependencyInjection/WindsorScopeFactory.cs:41]: Single line?

File: src/Castle.Windsor.Extensions.DependencyInjection/Scope/ExtensionContainerScope.cs
Comment (jonorossi) [inline src/Castle.Windsor.Extensions.DependencyInjection/Scope/ExtensionContainerScope.cs:30]: Field `rootScope` is unused.

File: src/Castle.Windsor.Extensions.DependencyInjection/Scope/ExtensionContainerScope.cs
Comment (jonorossi) [inline src/Castle.Windsor.Extensions.DependencyInjection/Scope/ExtensionContainerScope.cs:50]: `ArgumentNullException` for parameter `rootScope` doesn't make sense anymore since it isn't a parameter.

**Review by jonorossi — CHANGES_REQUESTED** [review]

File: src/Castle.Windsor.Extensions.DependencyInjection/Scope/ExtensionContainerScope.cs
Comment (generik0) [reply src/Castle.Windsor.Extensions.DependencyInjection/Scope/ExtensionContainerScope.cs:30]: it is used there
image

File: src/Castle.Windsor.Extensions.DependencyInjection/Scope/ExtensionContainerScope.cs
Comment (generik0) [reply src/Castle.Windsor.Extensions.DependencyInjection/Scope/ExtensionContainerScope.cs:50]: Root scope, when creating a scope must not be null. 
This couls happen if you ty to BeginScope, before "BeginRootScope" was called. I.e. if the user has there own WindsorServiceProviderFactory

File: src/Castle.Windsor.Extensions.DependencyInjection/WindsorScopeFactory.cs
Comment (generik0) [reply src/Castle.Windsor.Extensions.DependencyInjection/WindsorScopeFactory.cs:28]: image

File: src/Castle.Windsor.Extensions.DependencyInjection/WindsorScopeFactory.cs
Comment (generik0) [reply src/Castle.Windsor.Extensions.DependencyInjection/WindsorScopeFactory.cs]: Done

File: src/Castle.Windsor.Extensions.DependencyInjection/Scope/ExtensionContainerScope.cs
Comment (generik0) [reply src/Castle.Windsor.Extensions.DependencyInjection/Scope/ExtensionContainerScope.cs]: cleaned up now

File: src/Castle.Windsor.Extensions.DependencyInjection/WindsorScopeFactory.cs
Comment (generik0) [reply src/Castle.Windsor.Extensions.DependencyInjection/WindsorScopeFactory.cs]: cleaned up now

File: src/Castle.Windsor.Extensions.DependencyInjection/Scope/ExtensionContainerScope.cs
Comment (jonorossi) [inline src/Castle.Windsor.Extensions.DependencyInjection/Scope/ExtensionContainerScope.cs:48]: No need for a `$` since there is no interpolation.

I wasn't saying the exception isn't needed, but that an `ArgumentNullException` isn't the right exception:
> The exception that is thrown when a null reference (Nothing in Visual Basic) is passed to a method that does not accept it as a valid argument.

Right now you are passing that string as `paramName`:
```
public ArgumentNullException (string paramName);
```

Maybe `InvalidOperationException` makes more sense.

File: src/Castle.Windsor.Extensions.DependencyInjection/Scope/ExtensionContainerScope.cs
Comment (jonorossi) [inline src/Castle.Windsor.Extensions.DependencyInjection/Scope/ExtensionContainerScope.cs:32]: Double space

File: src/Castle.Windsor.Extensions.DependencyInjection/Scope/ExtensionContainerScopeAccessor.cs
Comment (jonorossi) [inline src/Castle.Windsor.Extensions.DependencyInjection/Scope/ExtensionContainerScopeAccessor.cs:26]: Double space

Haha, we've got an `InvalidOperationException` here for no scope available. Maybe should just have a static `CurrentOrThrow` property both can use?

File: src/Castle.Windsor.Extensions.DependencyInjection/WindsorScopedServiceProvider.cs
Comment (jonorossi) [inline src/Castle.Windsor.Extensions.DependencyInjection/WindsorScopedServiceProvider.cs:50]: You can just do `using (new ForcedScope(scope)) { ... }`, no need for the local.

File: src/Castle.Windsor.Extensions.DependencyInjection/WindsorScopedServiceProvider.cs
Comment (jonorossi) [inline src/Castle.Windsor.Extensions.DependencyInjection/WindsorScopedServiceProvider.cs:63]: I know this was there already before your pull request but it surprised me. Shouldn't it be the job of the person who created the container to dispose it, what if they wanted to continue using it. If there is a good reason for it having to work like this could we add a comment explaining that here or log a task to investigate.

**Review by jonorossi — CHANGES_REQUESTED** [review]
Sorry @generik0 for the back and forth on this pull request, I've got a few more review comments on the recent commits.

File: src/Castle.Windsor.Extensions.DependencyInjection/WindsorScopedServiceProvider.cs
Comment (generik0) [reply src/Castle.Windsor.Extensions.DependencyInjection/WindsorScopedServiceProvider.cs:63]: Yes i though the same thing when i refactored.
**I think** the reason is that the Provider is disposed with the host, and this is the reason for the container being disposed here.

File: src/Castle.Windsor.Extensions.DependencyInjection/Scope/ExtensionContainerScope.cs
Comment (generik0) [reply src/Castle.Windsor.Extensions.DependencyInjection/Scope/ExtensionContainerScope.cs:32]: > Sorry @generik0 for the back and forth on this pull request, I've got a few more review comments on the recent commits.

@jonorossi  I am happy for the reviews. I feel the change is getting better and better due to the good reviews. thankyou.

File: src/Castle.Windsor.Extensions.DependencyInjection/Scope/ExtensionContainerScope.cs
Comment (generik0) [reply src/Castle.Windsor.Extensions.DependencyInjection/Scope/ExtensionContainerScope.cs]: Improved

File: src/Castle.Windsor.Extensions.DependencyInjection/WindsorScopedServiceProvider.cs
Comment (generik0) [reply src/Castle.Windsor.Extensions.DependencyInjection/WindsorScopedServiceProvider.cs:63]: I.e. this is needed for the container to be dispoed when the Hosts stops. However i can do two thinks (maybe a seperate Issue):
- Add an option to the Provider Factory that disableds the automatic dispose of the container with the provider
- And, Or, make the disposable virtual to allow an override for the disposiing.
I will add an issue. Should not block this PR.

File: src/Castle.Windsor.Extensions.DependencyInjection/Scope/ExtensionContainerScopeAccessor.cs
Comment (generik0) [reply src/Castle.Windsor.Extensions.DependencyInjection/Scope/ExtensionContainerScopeAccessor.cs:26]: I can change, but may effect this in a funny way:
image

Comment (jonorossi) [reply src/Castle.Windsor.Extensions.DependencyInjection/Scope/ExtensionContainerScopeAccessor.cs:26]: Ah yer, probably easiest to just duplicate the throw statement for now.

File: src/Castle.Windsor.Extensions.DependencyInjection/WindsorScopedServiceProvider.cs
Comment (jonorossi) [reply src/Castle.Windsor.Extensions.DependencyInjection/WindsorScopedServiceProvider.cs:63]: Yer, just log a new issue to discuss it there to work out if it is a code change, code comments change or docs change since the docs don't mention you give ownership of the container over when you pass it.

File: src/Castle.Windsor.Extensions.DependencyInjection/Scope/ExtensionContainerScopeAccessor.cs
Comment (generik0) [reply src/Castle.Windsor.Extensions.DependencyInjection/Scope/ExtensionContainerScopeAccessor.cs]: I thnk the CurrentOrThrow is best, then you know exactly what is null. No Current, that will throw, no root, then this will throw. Please see update (still not my PR, i cannot resolve :-/ )

File: src/Castle.Windsor.Extensions.DependencyInjection/WindsorScopedServiceProvider.cs
Comment (generik0) [reply src/Castle.Windsor.Extensions.DependencyInjection/WindsorScopedServiceProvider.cs]: chagnes to using(_ = new ForcedScope(scope))
_ is discard, this is correct syntax i beleive

File: src/Castle.Windsor.Extensions.DependencyInjection/WindsorScopedServiceProvider.cs
Comment (generik0) [reply src/Castle.Windsor.Extensions.DependencyInjection/WindsorScopedServiceProvider.cs:63]: Please see change

File: src/Castle.Windsor.Extensions.DependencyInjection/Scope/ExtensionContainerScope.cs
Comment (generik0) [reply src/Castle.Windsor.Extensions.DependencyInjection/Scope/ExtensionContainerScope.cs]: fixed

File: src/Castle.Windsor.Extensions.DependencyInjection/Scope/ExtensionContainerScope.cs
Comment (generik0) [reply src/Castle.Windsor.Extensions.DependencyInjection/Scope/ExtensionContainerScope.cs]: No need for a $ since there is no interpolation. --> This was a R# recommendation. I will remove

File: src/Castle.Windsor.Extensions.DependencyInjection/WindsorScopedServiceProvider.cs
Comment (jonorossi) [inline src/Castle.Windsor.Extensions.DependencyInjection/WindsorScopedServiceProvider.cs:42]: > chagnes to using(_ = new ForcedScope(scope))
> _ is discard, this is correct syntax i beleive

Yes, `_` is the syntax for C# 7 discards. Happy either way, just didn't want to see an unused local.

**Review by jonorossi — CHANGES_REQUESTED** [review]

File: src/Castle.Windsor.Extensions.DependencyInjection/WindsorScopedServiceProvider.cs
Comment (generik0) [reply src/Castle.Windsor.Extensions.DependencyInjection/WindsorScopedServiceProvider.cs:42]: i removed the (**var** _) as you correctly recommended., thanks for that comment :-)

**Comment by generik0** [comment]
@jonorossi i cleaned up, and improved substitution and single responsibility a bit:
- ExtensionContainerScopeBase handles common code for ExtensionContainerScope and ExtensionContainerRootScope now
- ExtensionContainerScopeCache is now responsible for caching the scoping

I can revert if you do not feel i have SOLIDfied the Castle.Windsor.Extensions.DependencyInjection more

File: src/Castle.Windsor.Extensions.DependencyInjection/Scope/ExtensionContainerScopeBase.cs
Comment (jonorossi) [inline src/Castle.Windsor.Extensions.DependencyInjection/Scope/ExtensionContainerScopeBase.cs:25]: This constant should be marked `readonly` or defined as a constant.

File: src/Castle.Windsor.Extensions.DependencyInjection/Scope/ExtensionContainerScopeAccessor.cs
Comment (jonorossi) [inline src/Castle.Windsor.Extensions.DependencyInjection/Scope/ExtensionContainerScopeAccessor.cs:26]: `Current` already throws, no need for the check to throw here.

**Review by jonorossi — CHANGES_REQUESTED** [review]

File: src/Castle.Windsor.Extensions.DependencyInjection/Scope/ExtensionContainerScopeBase.cs
Comment (generik0) [reply src/Castle.Windsor.Extensions.DependencyInjection/Scope/ExtensionContainerScopeBase.cs:25]: done

File: src/Castle.Windsor.Extensions.DependencyInjection/Scope/ExtensionContainerScopeAccessor.cs
Comment (generik0) [reply src/Castle.Windsor.Extensions.DependencyInjection/Scope/ExtensionContainerScopeAccessor.cs:26]: done, and thanks

**Pull request merged**
